lib/base.js

```diff
@@ -1,138 +1,69 @@
-var ArrayElement = require('./primitives/array-element')(BaseElement, registry);
-var MemberElement = require('./primitives/member-element')(BaseElement, registry);
-var ObjectElement = require('./primitives/object-element')(BaseElement, ArrayElement, MemberElement);
-
-module.exports = {
-  BaseElement: BaseElement,
-  NullElement: NullElement,
-  StringElement: StringElement,
-  NumberElement: NumberElement,
-  BooleanElement: BooleanElement,
-  ArrayElement: ArrayElement,
-  MemberElement: MemberElement,
-  ObjectElement: ObjectElement,
-  ElementRegistry: ElementRegistry,
-  registry: registry
+  // Use dependency injection so we can split up into multiple files yet resolve
+  // any circuluar dependencies. BaseElement is also defined above in order to
+  // satisfy any linting issues.
+  var BaseElement = require('./primitives/base-element')(registry);
+
+  // The registry needs a base element to determine the base type of an element.
+  registry.BaseElement = BaseElement;
+
+  var NullElement = require('./primitives/null-element')(BaseElement);
+  var StringElement = require('./primitives/string-element')(BaseElement);
+  var NumberElement = require('./primitives/number-element')(BaseElement);
+  var BooleanElement = require('./primitives/boolean-element')(BaseElement);
+  var ArrayElement = require('./primitives/array-element')(BaseElement, registry);
+  var MemberElement = require('./primitives/member-element')(BaseElement, registry);
+  var ObjectElement = require('./primitives/object-element')(BaseElement, ArrayElement, MemberElement);
+
+  // Set up classes for default elements
+  registry
+    .register('null', NullElement)
+    .register('string', StringElement)
+    .register('number', NumberElement)
+    .register('boolean', BooleanElement)
+    .register('array', ArrayElement)
+    .register('object', ObjectElement)
+    .register('member', MemberElement);
+
+  // Add instance detection functions to convert existing objects into
+  // the corresponding refract elements.
+  registry
+    .detect(_.isNull, NullElement, false)
+    .detect(_.isString, StringElement, false)
+    .detect(_.isNumber, NumberElement, false)
+    .detect(_.isBoolean, BooleanElement, false)
+    .detect(_.isArray, ArrayElement, false)
+    .detect(_.isObject, ObjectElement, false);
+
+  return {
+    BaseElement: BaseElement,
+    NullElement: NullElement,
+    StringElement: StringElement,
+    NumberElement: NumberElement,
+    BooleanElement: BooleanElement,
+    ArrayElement: ArrayElement,
+    MemberElement: MemberElement,
+    ObjectElement: ObjectElement,
+    registry: registry,
+    convertToElement: function() {
+      return registry.toElement.apply(registry, arguments);
+    },
+    convertFromRefract: function() {
+      return registry.fromRefract.apply(registry, arguments);
+    },
+    convertFromCompactRefract: function() {
+      return registry.fromCompactRefract.apply(registry, arguments);
+    }
```

Comment:
Very true, but I can't figure out a way around it in my head. If you want to be able to have your own registry apart from the global one (which this PR does not have at the moment), you have to initialize all elements with that registry. It's the circular dependency issue. 

For example, the ArrayElement needs to have the registry to convert between types, but if you export it as a module on require, it will always have that one registry. Creating a custom registry will never work because ArrayElement doesn't know about it unless you store the new registry for each instance.

Storing a registry on each instance is a fundamental change, too, but may actually be the correct one given namespaces. But this still requires having some class that manages the scope (i.e. available registry) for a given document.